bbq/compiler/peephole_pass.go

```diff
@@ -0,0 +1,202 @@
+/*
+ * Cadence - The resource-oriented smart contract programming language
+ *
+ * Copyright Flow Foundation
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package compiler
+
+import (
+	"sort"
+
+	"github.com/onflow/cadence/bbq/opcode"
+	"github.com/onflow/cadence/interpreter"
+)
+
+type BytecodeShiftPair struct {
+	Offset int
+	Shift  int
+}
+
+type PeepholeOptimizer[E, T any] interface {
+	Optimize(code []E) []E
+}
+
+type PeepholeInstructionOptimizer[E, T any] struct {
+	PatternsByOpcode map[opcode.Opcode][]PeepholePattern
+	compiler         *Compiler[opcode.Instruction, interpreter.StaticType]
+	// bytecodeShifts is a list of where bytecode has been changed
+	// and how much the offset has changed. this is used to patch jumps
+	// after the peephole optimizations have been applied
+	bytecodeShifts []BytecodeShiftPair
+
+	// jumps is a list of offsets in the optimized bytecode that need to be patched
+	jumps []int
+}
+
+func NewPeepholeOptimizer[E, T any](compiler *Compiler[E, T]) PeepholeOptimizer[E, T] {
+	// this map is meant to improve pattern matching performance
+	patternsByOpcode := make(map[opcode.Opcode][]PeepholePattern)
+	for _, pattern := range AllPatterns {
+		if patternsByOpcode[pattern.Opcodes[0]] == nil {
+			patternsByOpcode[pattern.Opcodes[0]] = make([]PeepholePattern, 0)
+		}
+		patternsByOpcode[pattern.Opcodes[0]] = append(patternsByOpcode[pattern.Opcodes[0]], pattern)
+	}
```

Comment:
Is the resulting map (`patternsByOpcode`), or any of the values (`[]PeepholePattern` slices) ever mutated after construction? From what I can see it is not. Does the map differ for each optimizer instance? 

If not, we can construct this lookup table in a global variable and construct it at package initialization time by moving this code into an [`init()` function](https://go.dev/doc/effective_go#init)